extensions/keycloak-sms-authenticator/src/main/java/com/inventage/keycloak/sms/credential/SmsCredentialProvider.java

```diff
@@ -29,30 +35,65 @@ public CredentialModel createCredential(RealmModel realm, UserModel user, SmsCre
             credentialModel.setCreatedDate(Time.currentTimeMillis());
         }
 
-        return allowMultipleCredentials() ? addCredential(user, credentialModel) : updateCredential(realm, user, credentialModel);
+        return allowMultipleCredentials() ? addCredential(realm, user, credentialModel) : updateCredential(realm, user, credentialModel);
     }
 
     private boolean allowMultipleCredentials() {
         return true;
     }
 
-    private CredentialModel addCredential(UserModel user, SmsCredentialModel credentialModel) {
-        return user.credentialManager().createStoredCredential(credentialModel);
+    private CredentialModel addCredential(RealmModel realm, UserModel user, SmsCredentialModel newCredential) {
+        if (samePhoneNumberAlreadyConfigured(user, newCredential)) {
+            LOGGER.debugf("addCredential: User already has a credential with same phone number. So old credential will be updated.");
+            final CredentialModel oldCredential = getSmsCredentialModel(realm, user);
+            return updateAvailableCredential(user, oldCredential, newCredential);
+        }
+
+        LOGGER.debugf("addCredential: New SMS credential will be added.");
+        return user.credentialManager().createStoredCredential(newCredential);
+    }
+
+    private boolean samePhoneNumberAlreadyConfigured(UserModel user, SmsCredentialModel credentialModel) {
+        Optional<CredentialModel> smsCredentialModel = getCredentialModelWithSamePhoneNumber(user, credentialModel);
+
+        return smsCredentialModel.isPresent();
+    }
+
+    private Optional<CredentialModel> getCredentialModelWithSamePhoneNumber(UserModel user, SmsCredentialModel credentialModel) {
+        return user.credentialManager().getStoredCredentialsByTypeStream(getType())
+            .filter
+                (
+                    cred -> {
+                        try {
+                            final SmsCredentialData smsCredentialData;
+                            smsCredentialData = JsonSerialization.readValue(cred.getCredentialData(), SmsCredentialData.class);
```

Comment:
I meant as a one-liner.

```suggestion
                            final SmsCredentialData smsCredentialData = JsonSerialization.readValue(cred.getCredentialData(), SmsCredentialData.class);
```